google/cloud/storage/tests/object_integration_test.cc

```diff
@@ -347,6 +347,29 @@ TEST_F(ObjectIntegrationTest, InsertWithMD5) {
   client.DeleteObject(bucket_name, object_name);
 }
 
+TEST_F(ObjectIntegrationTest, InsertWithComputedMD5) {
+  Client client;
+  auto bucket_name = ObjectTestEnvironment::bucket_name();
+  auto object_name = MakeRandomObjectName();
+
+  std::string expected = LoremIpsum();
+
+  // Create the object, but only if it does not exist already.
+  ObjectMetadata meta =
+      client.InsertObject(bucket_name, object_name, expected,
+                          IfGenerationMatch(0),
+                          MD5HashValue(ComputeMD5Hash(expected)));
+  EXPECT_EQ(object_name, meta.name());
+  EXPECT_EQ(bucket_name, meta.bucket());
+
+  // Create a iostream to read the object back.
+  auto stream = client.ReadObject(bucket_name, object_name);
```

Comment:
Hmm.  In `InsertObjectMediaXml()`, we could only construct an `ObjectMetadata` using the bucket name and object name.  We *could* also be checking the `x-goog-hash` response headers and populating the checksum fields in the returned `ObjectMetadata` there, which would prevent us from having to issue a separate read call here.  But that's probably out of scope for this CL; just a consideration.

Reply:
Ack.  Agree.